src/server/req_runjob.c

```diff
@@ -936,7 +936,9 @@ svr_startjob(job *pjob, struct batch_request *preq)
 			return (PBSE_SYSTEM);
 
 	/* clear Exit_status which may have been set in a hook and requeued */
-	clear_attr(get_jattr(pjob, JOB_ATR_exit_status), &job_attr_def[(int) JOB_ATR_exit_status]);
+	if (job_delete_attr(pjob, JOB_ATR_exit_status)) {
```

Comment:
Do we need an explicit write to the database? can you set the exit status to zero which will in turn get written into the db in the next db operation?

Reply:
Yes, we need the explicit deletion. Setting exit_status to zero will use the SQL `attributes = attributes || hstore($number::text[])` in the next db operation and that is the problem. As mentioned in this PR's opening post, the `||` - OR operation will not delete the attribute from db. It will only update and join.

We explicitly need to call `pbs_db_delete_attr_obj()` - that is the way to remove an attribute from db.

As you can see, there was only one place the `pbs_db_delete_attr_obj()` is called and that is `req_manager.c:mgr_unset_attr()` before this PR. The `pbs_db_delete_attr_obj()` is imminent once the attribute is removed from memory. So we need a similar solution for the job attributes - if we want to remove it.

Reply:
I was not clear in the previous comment. What I meant is, instead of removing or clearing the attribute, can we set the exit_status to zero? Since it's an update, it will still get written to the database in the subsequent operation. Will the job be considered for a rerun in that case?

I am concerned about the overhead that a single attribute write can cause if we add more such writes in the future. How about unsetting the attribute flag? Such attributes are not loaded on read.

Reply:
Clearing an attribute should still work with an update, as opposed to freeing/deleting it. What am I missing?

Reply:
Hmm, I understand now. Well, the troublemaker is the exit_status -18 (JOB_EXEC_HOOK_RERUN). The job must have the exit_status set to -18 to invoke this bug. So I tested setting the job Exit_status to zero here and it seems to work. Of course, I tried to set the zero only if some value was set before. I am afraid it could break some cases with subjobs and I also think it could have unintended consequences in cpy_stdfile() but I was not able to invoke any related bug with this solution. Seems working.

Honestly, I did not consider this solution before because I do not feel it is the right way to fix this issue. I don't like to set exit_status to any number on the job start. Exit_status = 0 indicates the job run and finished with success. IMO, It does not make sense to set it to zero because we do not want to write to the database.

Maybe, we could set the exit_status to a new value JOB_EXEC_UNKNOWN if the exit_status is set on job start... IMO that makes a bit more sense but I am still not a fan. I still think deleting/clearing the exit_status attribute on the rerun/job start is the right thing to do. Once the job is running, the previous exit_state is definitely invalid, so why bother a user with any value?

I understand your concern about the overhead and I agree the `pbs_db_delete_attr_obj` should be used with caution. IMO the -18 rerun should be relatively a rare occasion, so we are hopefully safe with this use. 

Setting Exit_status by hooks before the job start is just a communication helper. It is a helper for how to pass the result of the hook run to the server. So I have one more solution in my mind: Maybe we could add a new attribute. Not sure how to name it but something like `hook_status` and do not bother the exit_status with hook exit codes... ?

I am not sure how to resolve. Honestly, I still like the clearing/deleting as the best/right way to fix this issue. Let me know if I should pursue the `JOB_EXEC_UNKNOWN` or the `hook_status` solution.

As I understand the flag ATR_VFLAG_SET, it is set when an attribute is set and the flag itself is not saved in the database. It means the attribute exit_status will be read on the server restart and the ATR_VFLAG_SET will be set again. IMO, this is not a solution.

Reply:
Considering the options and their implications, clearing or deleting the exit_status on rerun/job start seems to be the most logical and clean approach. This solution should be viable as long as the performance impact remains minimal, given that the -18 rerun is a relatively rare occurrence.